Subject: f2fs: make the debugging files work only for the requested superblock

From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Right now, each debug file shows the information for all superblocks,
not just the one being referenced.  This patch fixes that.


---
 fs/f2fs/debug.c |   24 ++++++------------------
 1 file changed, 6 insertions(+), 18 deletions(-)

--- a/fs/f2fs/debug.c
+++ b/fs/f2fs/debug.c
@@ -270,20 +270,10 @@ static const struct file_operations sit_
 
 static int mem_show(struct seq_file *s, void *v)
 {
-	struct f2fs_gc_info *gc_i, *next;
-	struct f2fs_stat_info *si;
+	struct f2fs_sb_info *sbi = v;
+	unsigned npages;
+	unsigned base_mem = 0, cache_mem = 0;
 
-	list_for_each_entry_safe(gc_i, next, &f2fs_stat_list, stat_list) {
-		struct f2fs_sb_info *sbi = gc_i->stat_info->sbi;
-		unsigned npages;
-		unsigned base_mem = 0, cache_mem = 0;
-
-		si = gc_i->stat_info;
-		mutex_lock(&si->stat_list);
-		if (!si->sbi) {
-			mutex_unlock(&si->stat_list);
-			continue;
-		}
 		base_mem += sizeof(struct f2fs_sb_info) + sbi->sb->s_blocksize;
 		base_mem += 2 * sizeof(struct f2fs_inode_info);
 		base_mem += sizeof(*sbi->ckpt);
@@ -336,8 +326,6 @@ static int mem_show(struct seq_file *s,
 		seq_printf(s, "%u KB = static: %u + cached: %u\n",
 			   (base_mem + cache_mem) >> 10,
 			   base_mem >> 10, cache_mem >> 10);
-		mutex_unlock(&si->stat_list);
-	}
 	return 0;
 }
 
@@ -407,15 +395,15 @@ int f2fs_stat_init(struct super_block *s
 		return -EINVAL;
 
 	if (!debugfs_create_file("f2fs_stat", S_IRUGO, sbi->s_debug,
-				 NULL, &stat_fops))
+				 sbi, &stat_fops))
 		goto failed;
 
 	if (!debugfs_create_file("f2fs_sit_stat", S_IRUGO, sbi->s_debug,
-				 NULL, &sit_fops))
+				 sbi, &sit_fops))
 		goto failed;
 
 	if (!debugfs_create_file("f2fs_mem_stat", S_IRUGO, sbi->s_debug,
-				 NULL, &mem_fops))
+				 sbi, &mem_fops))
 		goto failed;
 
 	return 0;
